### PR TITLE
SOF-7161: move build schemas/examples to esse class

### PR DESCRIPTION
**Pull request by k0stik**

### CODE REVIEW

File: tests/js/JSONSchemasInterface.tests.ts [hunk 1]
```diff
@@ -23,7 +23,9 @@ describe("JSONSchemasInterface", () => {
         expect(schema?.properties?.inSet).to.be.an("object");
 
         if (
+            // @ts-ignore
```
Comment (timurbazhirov) [inline tests/js/JSONSchemasInterface.tests.ts]: Maybe just fail if schema is undefined instead or add it to if?

File: tests/js/validate.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import path from "path";
+
+import { JSONSchemasInterface } from "../../src/js/esse/JSONSchemasInterface";
+import { walkDirSync } from "../../src/js/scripts/utils";
+
+const examplesPath = path.resolve("./lib/js/example");
+const schemasPath = path.resolve("./lib/js/schema");
+
+describe("validate all examples", () => {
+    console.log("======");
```
Comment (timurbazhirov) [inline tests/js/validate.ts]: Remove?

File: tests/js/validate.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/* eslint-disable no-unused-expressions */
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (timurbazhirov) [inline tests/js/validate.ts]: Should we remove these?

File: tests/js/validate.ts (hunk 2 above)
Comment (k0stik) [reply tests/js/validate.ts]: Done

File: tests/js/JSONSchemasInterface.tests.ts (hunk 1 above)
Comment (k0stik) [reply tests/js/JSONSchemasInterface.tests.ts]: Right. Removed ifs

File: tests/js/validate.ts (hunk 3 above)
Comment (k0stik) [reply tests/js/validate.ts]: Done

File: src/js/esse/index.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    propertiesManifestDir?: string;
+}
+
+const DEFAULT_CONFIG = {
+    schemasDir: SCHEMAS_DIR,
+    examplesDir: EXAMPLES_DIR,
+    propertiesManifestDir: PROPERTIES_MANIFEST_PATH,
+};
+
+export class ESSE implements EsseConfig {
```
Comment (timurbazhirov) [inline src/js/esse/index.ts]: ESSE -> JSONSchemasGenerator
EsseConfig -> JSONSchemasGeneratorConfig

File: src/js/scripts/setSchemaIds.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import fs from "fs";
+import path from "path";
+
+import { walkDirSync } from "./utils";
+
+const SCHEMA_DIR = "../../../schema/";
+
+walkDirSync(SCHEMA_DIR, (filePath) => {
```
Comment (timurbazhirov) [inline src/js/scripts/setSchemaIds.ts]: Let's explain what this script does in a comment

File: src/js/scripts/utils.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable no-await-in-loop */
+/* eslint-disable no-restricted-syntax */
+import fs from "fs";
+import fsPromises from "fs/promises";
+import path from "path";
+
+export async function walkDir(dir: string, callback: (itemPath: string) => void | Promise<void>) {
```
Comment (timurbazhirov) [inline src/js/scripts/utils.ts]: Should we have a folder utils and isolate all utils that do not depend on anything there?

`js/utils/filesystem.ts`

File: src/js/esse/JSONSchemasInterface.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            return;
+        }
+        const schema = JSON.parse(fs.readFileSync(filePath).toString());
+        schemas.push(schema);
+    });
+
+    return schemas;
+}
+
+export class JSONSchemasInterface {
```
Comment (timurbazhirov) [inline src/js/esse/JSONSchemasInterface.ts]: Should inherit from Client

File: src/js/esse/JSONSchemasInterfaceClient.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import { SchemaObject } from "ajv";
+import fs from "fs";
+import path from "path";
+
+import { walkDirSync } from "../scripts/utils";
+import { JSONInterfaceQuery } from "./types";
+
+export function readSchemaFolderSync(folderPath: string) {
```
Comment (timurbazhirov) [inline src/js/esse/JSONSchemasInterfaceClient.ts]: Remove

File: src/js/esse/JSONSchemasInterfaceClient.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+export default class JSONSchemasInterfaceClient {
+    static schemasCache = new Map<string, SchemaObject>();
+
+    static addSchemaObject(schema: SchemaObject) {
+        if (schema.$id) {
+            this.schemasCache.set(schema.$id, schema);
+        }
+    }
+
+    static schemaById(schemaId: string) {
```
Comment (timurbazhirov) [inline src/js/esse/JSONSchemasInterfaceClient.ts]: getSchemaById

File: src/js/esse/JSONSchemasInterfaceClient.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        schemas.push(schema);
+    });
+
+    return schemas;
+}
+
+export default class JSONSchemasInterfaceClient {
+    static schemasCache = new Map<string, SchemaObject>();
+
+    static addSchemaObject(schema: SchemaObject) {
```
Comment (timurbazhirov) [inline src/js/esse/JSONSchemasInterfaceClient.ts]: addSchema

File: src/js/esse/JSONSchemasInterfaceClient.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            return;
+        }
+        const schema = JSON.parse(fs.readFileSync(filePath).toString());
+        schemas.push(schema);
+    });
+
+    return schemas;
+}
+
+export default class JSONSchemasInterfaceClient {
```
Comment (timurbazhirov) [inline src/js/esse/JSONSchemasInterfaceClient.ts]: Functional:

1. Building and knowing which schemas we have (SERVER ONLY at the transpile time)

	Build Schemas
	Create Index (to know what's available)


2. Managing the in-memory set of schemas and returning schema to requester by ID

	Create in-memory cache with all schemas
	Expose the Index in memory

	Methods:

	- addSchema
	- getSchemaById

3. Extending the default list with additional schemas

	Add a set of schemas to in-memory cache by passing a folder

File: src/js/scripts/setSchemaIds.ts (hunk 5 above)
Comment (k0stik) [reply src/js/scripts/setSchemaIds.ts]: Done

File: src/js/scripts/utils.ts (hunk 6 above)
Comment (k0stik) [reply src/js/scripts/utils.ts]: Done

File: src/js/esse/JSONSchemasInterface.ts (hunk 7 above)
Comment (k0stik) [reply src/js/esse/JSONSchemasInterface.ts]: Done

File: src/js/esse/JSONSchemasInterfaceClient.ts (hunk 8 above)
Comment (k0stik) [reply src/js/esse/JSONSchemasInterfaceClient.ts]: Done

File: src/js/esse/JSONSchemasInterfaceClient.ts (hunk 9 above)
Comment (k0stik) [reply src/js/esse/JSONSchemasInterfaceClient.ts]: Done

File: src/js/esse/JSONSchemasInterfaceClient.ts (hunk 10 above)
Comment (k0stik) [reply src/js/esse/JSONSchemasInterfaceClient.ts]: Done

File: src/js/esse/index.ts (hunk 4 above)
Comment (k0stik) [reply src/js/esse/index.ts]: Done

**Comment by k0stik** [comment]
See https://github.com/Exabyte-io/esse/pull/299

**Pull request closed without merging**